WalletWasabi/Services/WalletService.cs

```diff
@@ -722,6 +872,43 @@ public async Task<Block> GetOrDownloadBlockAsync(uint256 hash, CancellationToken
 			return block;
 		}
 
+		private static bool TryFetchBlockFromDisk(string blocksFolderPath)
+		{
+			var blocksFolder = new DirectoryInfo(blocksFolderPath);
+			var totalCount = 0;
+			var noChange = true;
+			foreach (var blkFile in blocksFolder
+				.EnumerateFiles("blk*.dat", new EnumerationOptions() { IgnoreInaccessible = true, RecurseSubdirectories = false })
+				.OrderByDescending(x => int.Parse(x.Name.Split(new[] { "blk", ".dat" }, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault())))
+			{
+				// If nothing changed, then don't bother updating.
+				if (noChange)
+				{
+					if (blkFile.Length == LastBlkSize && blkFile.Name == LastBlkName)
+					{
+						break;
+					}
+					LastBlkSize = blkFile.Length;
+					LastBlkName = blkFile.Name;
+					LastBlocksFromCore.Clear();
+					noChange = false;
+				}
+
+				foreach (var storedBlock in StoredBlock.EnumerateFile(blkFile.FullName))
+				{
+					totalCount++;
+					LastBlocksFromCore.Add(storedBlock.Item.GetHash(), storedBlock.Item);
```

Comment:
Even in the case that the `blkxxx.dat` changed, it is highly probable that it still contains mostly the same set of blocks that before. In such case this line will trow a `DuplucatedKeyException`.